Multiple Patient Addresses - Update Appearance & Styling

psoas opened the pull request:
Fixes:  Demographics Display Readability #5548 

Also:
Fixes Issue where demographic styles were not being implemented in Demographics Display Mode.
Overhauls how "Additional Addresses" are displayed in Demographics, both in Display & Edit Mode.
Adds a triangle to open & close input form in Demographics Edit Mode.
Makes Additional Addresses more responsive - but probably more will need to be done.

bradymiller commented:
Can test this PR at the following Up For Grabs demo:
https://www.open-emr.org/wiki/index.php/Development_Demo#Epsilon_-_Up_For_Grabs_Demo

bradymiller commented:
Seeing a lot of wasted whitespace to left on my testing:
![image](https://user-images.githubusercontent.com/278968/177005778-5ec8bf57-3877-46ac-9cb3-573c8e9aca0a.png)
![image](https://user-images.githubusercontent.com/278968/177005808-39986a52-12f8-44d5-a363-7e173e7e563f.png)

Also seems odd to have a global to change behavior of just one form, which will be different than all the other forms; for example, the dem and insurance layouts are different in 2nd screenshot. Rather than making bunch of globals for these things prob best to settle on a standard.

sjpadgett commented on library/templates/address_list_display.php:
```diff
@@ -15,76 +15,159 @@
-        <tr>
-            <td class='text-addresses-full-address'></td>
-        </tr>
+    <div class="table_text_addresses">
+        <div class ="text_addresses header label_custom row">
+            <?php echo xlt("Additional Addresses"); ?>
+        </div>
+        <div class="d-none no_addresses row">
+                <span class="text data"><?php echo xlt("NONE"); ?></span>
+        </div>
```
nice to see the table stuff gone. However, since view is one of the last vestige of using tables in engine, I'd recommend we start refactoring to bootstrap and divs. Then use BS col grid same as edit that Rod had done. Meanwhile breaking datatypes out as separate twig templates.

psoas replied:
@sjpadgett
Sounds nice.  Refactoring code from table to div is probably well beyond the scope of what I am doing here.  

Not sure about twig templates if people won't use twig feature (i.e. calendar code is supposed to be twig templates but there is a lot of php code inserted in it using few twig templating feature.

sjpadgett commented on interface/themes/core/patient/demographics.scss:
```diff
@@ -34,3 +35,116 @@
+}
+
+#DEM .text_addresses.header {
+  background: #e9ecef;
```
colors in scss should use theme color override variables.

psoas replied:
@sjpadgett
Yes, I should have used the variable $gray-400 in the scss code.

However, I now moved this to inline per your and Brady's suggestion.  Bootstrap has limited colors, and none were appropriate.  

gray-400 does not work in css.  How do I get named colors?  Do I need to add scss code to add named colors of gray-100 to gray-900?  Can I?  How and where do I do this?  

I also tried var(--gray-400) in the css styling but this did not work.  Why not?  From what I read, it should have.

sjpadgett replied:
This is defining the var to use elsewhere.

sjpadgett commented on interface/themes/core/patient/demographics.scss:
```diff
@@ -34,3 +35,116 @@
+}
+
+#DEM .label_custom {
+  text-align: right;
```
This is what makes form look weird. Try auto however due to the way LBF renders using table columns I doubt you'll be able to get this to look the way you hope i.e data not proportional in view. I think until we get rid of the tables in this view we'll go in circles:)

sjpadgett commented:
Hi @psoas You where correct about bug in span entire row option. See: PR #5584 thanks.

sjpadgett commented on interface/patient_file/summary/demographics.php:
```diff
@@ -851,6 +851,12 @@ function setMyPatient() {
             border-radius: 0;
         }
 
+        <?php if (!$GLOBALS['right_justify_labels_demographics']) { ?> 
+        #DEM .label_custom {
+            text-align: left !important;
+        }
+
+        <?php } ?>
```
I'd remove all these styles and check my PR

psoas replied:
@sjpadgett
I know that you want to decide on a standard, but some people like one way, other people another.  Giving an option is easy.  Why not have an option?  However, it needs to be fixed to also include insurance and perhaps other areas.  You also see that in the code where I inserted this, someone else did the say thing for something else.  Will you remove that also.

sjpadgett replied:
You can leave what is already there however, you may be able to get by with just BS style. As for the global well that's for Brady because I don't mind too much.
Try to remember though that we are going to refactor a lot of this soon. I almost did it my last PR...

bradymiller replied:
hi @psoas , Can live with a new global (consider it your one "get yourself out of globals jail" card :) ) as long as don't get all that empty left whitespace I noted in the above screenshots.
Also, if you have new code changes, then would rec pushing it to this PR.

sjpadgett commented on interface/patient_file/summary/demographics.php:
```diff
@@ -899,7 +905,7 @@ function setMyPatient() {
 </head>
 
-<body class="mt-1 patient-demographic bg-light">
+<body id="DEM" class="mt-1 patient-demographic bg-light">
```
Not really a good idea to add the id to body. My PR where I added styles to the tabs.scss is prob the better place to do styling in demographic view.

psoas replied:
@sjpadgett
This is a tricky issue.  

There is a decent amount of styling that already exists in the code base that refers to id of "DEM".  Should I move all of that styling to tabs.scss and not have it depend on the id "DEM".  This might be my preferred choice, but I don't know if it would mess up other areas.  It would be easiest for me to do this after your PR is merged.  How long will that be?

OR
I can create a div nested just inside body with both id="DEM" class="mt-1 patient-demographic bg-light".  I would think that this would be redundant since it could have just been done to body.

Personally, I think that DEM should have been a class and not an ID in the first place.  But to do that change would require changing even more code, which I want to avoid.

Also, if it is bad to give the tag body an ID, I would think that it would be just as bad to give it a class.  Is it?

I tried to search internet about using ID's and Classes for body tag.  I could not find anything that said that either was bad.

sjpadgett replied:
I merged my PR and also cherry picked it back v7.0 release.
Ids are unique to their scope where a class may be used more than once.

It's not bad to give the body tab an id but it's the DEM id I don't trust.

sjpadgett commented on interface/themes/core/patient/demographics.scss:
```diff
@@ -34,3 +35,116 @@
+
+/* Demographics - Previous Addresses - address_list_display.php */
+
+#DEM .table_text_addresses {
+  border-bottom: 2px solid #e9ecef;
+  margin-top: 1em;
+}
+
+#DEM .text_addresses.header {
```
I'd bootstrap style you addresses widget so you're not tied to the DEM id in case the widget is used elsewhere which the way it is shaping up, is very likely.

psoas replied:
@sjpadgett
Based on weekly conference call discussion, I addressed this.  I got rid of all of the styling that I put in demographics.css and used bootstrap classes & inline classes.

sjpadgett commented:
@psoas I think this is what you want for view. There isn't a need to modify any existing style sheets.
![image](https://user-images.githubusercontent.com/1859985/177372065-e5f7aa99-8938-4d4c-9f1e-fa91868fa7ef.png)

I only did the display template and I think you should prob get rid of alternative global display mode. Just not needed and again this all will be refactored soon. I emailed `library/templates/address_list_display.php`

psoas commented:
Hi @sjpadgett @bradymiller
I already finished my code modifications using bootstrap classes so that i don't need to modify existing style sheets.  

I just had some adjustments due to tabs.scss changes.

Now, I need to look at jerry's suggestion.

Brady, I know that you were in a rush.  Do you think that i might have till fri?

sjpadgett commented:
@psoas With the minor mod I made to your display template and I pulled down your address list form without any style changes to demographics edit/view inline styles and looks good to me. So in the end the only two files needing changes are the template files.
So I'll leave with you and Brady while I go meddle elsewhere!:)

psoas commented:
@sjpadgett, @bradymiller 

Latest Updates:

--Got rid of CSS styling in demographics.scss.  Moved styling to the php files, using combination of bootstrap classes, html styling tags, and inline styles, as per Brady's & Jerry's suggestion.  Everything displays well on firefox, hopefully it will be the same on other browsers.

--Updated demographics_full.php so that Insurance is also right justified when my global option is checked, as per Brady's suggestion.

--Fixed label display and alignment issues in both demographics screens and insurance.  These were always misaligned, just more noticeable with right justification of labels.  See attached screenshot to show insurance display issue that I fixed.

--Minor fix to tabs.scss so that styling specified works in both display and edit modes.  Previously, in only worked in display mode.

A lot of work for minor issues, but learning a lot.  Hopefully, you will accept changes.

psoas commented:
<img width="1150" alt="Screenshot 2022-07-10 - Insurance - Edit" src="https://user-images.githubusercontent.com/121007/178159996-12aa364c-d122-45c9-9cb0-5d7bfffdedb7.png">

sjpadgett commented on library/templates/address_list_form.php:
```diff
@@ -39,208 +39,223 @@
-                    ], '');
-                    ?>
-                </div>
+            <div class="fas fa-fw fa-trash-alt text-danger btn-delete-address" style="text-align: center; flex: 0 0 2em; line-height: 1.5;">
```
Need to give this tag a role so we get a cursor pointer for the fake button
`<div class="fas fa-fw fa-trash-alt text-danger btn-delete-address" role="button" style="text-align: center; flex: 0 0 2em; line-height: 1.5;">`
This could easily be done as below in keeping with themes
`<div class="btn btn-delete text-danger p-0 m-0 btn-delete-address">`
![image](https://user-images.githubusercontent.com/1859985/178329288-6ed23187-81e0-400d-8fa9-aca50337ec07.png)

I don't understand why we need inline styles for alignments. For one I question why we are vertical aligning to top by default.(Not on you David). All these alignments can be done in BS classes and should have been sorted out when we refactored demographic.

Soon I'll be going through all this when I refactor out table views.

sjpadgett reviewed (changes requested):
David I've made some comments where some are necessary changes while others could be argued.
I'm going to wait until @bradymiller takes a look and you make some adjustment until I pull down to test.
Otherwise you're doing great and thanks for the contribution...

psoas commented on library/templates/address_list_display.php:
```diff
@@ -15,78 +15,163 @@
+
+<div id="<?php echo attr($table_id); ?>" class="row mt-3">
+    <div class="table_text_addresses col-12">
+        <div class ="text_addresses_header label_custom pl-1" style="text-align: left !important; background: rgba(34, 34, 34, 0.1);">
```
@sjpadgett 
I want the header color to match the color of the lines below it (specified by HR tag).  bg-light is much too light for the header and does not match HR color.  Ironically, the default color for bootstrap table header and lines is the perfect shade of gray, and it matches the default line color of HR tag, but there is no named color for it.  What named color can I use to match color of the line in HR tag?  bg-light is not acceptable to me.  Its too light and you can barely see it.  bg-secondary is too dark.

sjpadgett replied:
I will think about this and see what is available in BS 5 then perhaps we'll create a class for your custom color.

sjpadgett replied:
@psoas I agree. I like the color and have myself looked for a similar color in the past for the same reason.
What really should happen is the new color should be added to our themes color palette but I don't have the energy.
Just leave as is

sjpadgett commented:
@psoas 
All the default color variables for BS are in interface/themes/colors/utilities/default_variables.scss 
This sheet gets compiled in BS base as variables to be used throughout. So to add new color this is one place to add(there are other places too). Since gray shades are defined already we can compile one where needed. In this case I chose navmenu.scss
So add these to that sheet:
```

  .bg-gray-2 {
    background-color: $gray-200;
  }

  .bg-gray-3 {
    background-color: $gray-300;
  }

  .bg-gray-4 {
    background-color: $gray-400;
  }

  .bg-gray-5 {
    background-color: $gray-500;
  }
``` 

Then compile npm run build and add the class to your header e.g. bg-gray-3
![image](https://user-images.githubusercontent.com/1859985/178497573-9931f374-4983-4a87-aca8-9af77d0ed7dd.png)

Guess I found the energy!:)

sjpadgett commented:
Beats me! We need to get this in because we want to release v7. Need to fix my other issues and just leave header color as is.

psoas commented:
I am hoping to have time to finish and complete all suggested changes by tomorrow (WED).  I will leave the issue of figuring out how the color for HR is defined in the scss files for later and for now just use one of the gray shades such as gray-400, defining the bg classes as you suggested.

psoas commented:
@bradymiller 
Hi.  I am almost finished with Padgett's suggested changes to my code but I am really busy with my “day job” that I might not be able to submit them till friday or Saturday.  Will that be too late?

bradymiller commented:
hi @psoas , just submit whatever you got now (if nothing yet, just let us know) and we'll see if one of us can get it done before then (goal is release over weekend, but I generally build the package/dockers a couple days before that for sanity checking).

psoas commented:
@bradymiller @sjpadgett 
Finished.  

I could not get rid of all of the inline styling.  Everytime I messed with one things, all the alignments became off.

I studied the themes styling code and now fully understand how it works.  I added the gray shades to the themes array in variables_custom.  This allows bootstrap to automatically add those shades to the bg- and border- classes.  It is not the way Jerry suggested, but it is more instep with how bootstrap works.  This also allowed me to use classes to match the HR color with the header color using "named" colors.

sjpadgett commented on interface/themes/oe-styles/style_light.scss:
```diff
@@ -1,9 +1,12 @@
 $close-text-shadow: 0 1 0 $black;
 $gray: #cdcdcd;
+
+@import "../default-variables";
```
This worries me a bit! It was patched out for a reason which I don't know. Pathing probably.
Light theme is kind of a special theme that broke with our older color themes so it renders differently in many ways.

I'd say you're correct but I still worry. Need to test well.

sjpadgett reviewed (approved):
Will merge in couple hours

sjpadgett commented on library/globals.inc.php:
```diff
@@ -421,6 +421,13 @@ function gblTimeZones()
+        'right_justify_labels_demographics' => array(
+            xl('Right Justify Labels in Demographics'),
+            'bool',                           // data type
+            '1',                              // default = true
```
DO NOT want this on by default.

psoas commented:
@bradymiller @sjpadgett 
Is there any way to put directions on the upgrade that a "regular" person will see that states that they need to run "npm run build".  This would be for the non-professional IT people/programmers using OpenEMR.  Is there any way a statement can come up when someone runs "http://localhost/openemr/sql_upgrade.php"?

sjpadgett commented:
upgrades are already built when we release so no reason for building by user. devs know to build working from master.

sjpadgett commented:
I've cherry picked to our v7.0.0 release branch.

psoas commented:
@sjpadgett 
Oh, so does that mean the upgrade script runs "npm run build"?  Is it able to?  Is it in "http://localhost/openemr/sql_upgrade.php"?  I could swear with upgrade I needed to still run "npm run build".  But, my recollection might be fuzzy.

sjpadgett commented:
No we build the release and include result from build in release zip/tar. Upgrade sql is for database changes.

bradymiller commented:
![2DV](https://user-images.githubusercontent.com/278968/179147826-3c39baa7-40c3-4a65-bb55-37c9d049533b.gif)
